Step3 : 자동차 경주 (두번째 리뷰 요청)

Mo-bile opened the pull request:
안녕하세요? 👍 
이전에 드린 PR에 커멘트 남긴 뒤 리뷰 요청 드리려고 했습니다.
그런데 실수로 본 [PR](https://github.com/next-step/java-racingcar/pull/6189) 과 관련된 origin branch 를 삭제를 해서 PR 이 closed 가 됐네요... 😂 이 부분 양해부탁드리겠습니다. 저도 당황스럽네요... close 된 PR re-open 하는 방식을 시도했지만 'The step3 branch was force-pushed or recreated.' 라면서 불가능 하다고 합니다... 최종 merge 를 위해 다시 PR을 열어야할까요?

두번째 PR은 이전에 요청주신 '핵심 비지니스 로직과 UI 로직을 분리하는 연습' 에 집중하였습니다.
특히 출력부분을 비즈니스 로직에서 분리하는데 새로운 model 이 필요하고 클래스 분리까지 고민하게 되었습니다.

## 질문
### 1. 역할과 책임에 집중하여 클래스를 분리하려고 하는데 이부분이 쉽지 않았습니다. 어떤 기준으로 나누는게 좋을까요?
클래스 분리에 책임단위라는 느낌이 어려워서 개인적으로 '맥락(context)' 가 달리한다면 나눈다는 기준을 두었습니다.
왜나하면 각각의 역할과 책임은 맥락이 다르기 때문입니다. 어떻게 생각하실까요?

그럼에도 불구하고 맥락에 따라 나누려고하는데 클래스를 보는 맥락을 어떤관점으로 보느냐에 따라 달리 볼수있겠구나 싶어습니다.
예로들어서 현실의 세계에서도 `경찰이 보는 사람`, `의사가 보는 사람`, `은행원이 보는 사람` 모두 다른것 처럼요
개인적으로 알아보니 개발에서는 domain, actor, 기능 관점으로 나눈다고 하는데 혹시 어떤 기준으로 나누시는지요?

### 2. 과도한 추상화를 멈추는 기준이 있을까요?
무언가 클래스 분리, 메서드 분리를 하려고하는데 욕심이 과하다 보니 과도한 추상화를 하고있다고 생각이 들어서 몇번 멈추고 되돌렸습니다.
롤백하면서 내다보낸 시간이 아쉬웠습니다. 혹시 어떤기준으로 고민하면 좋을까요?

주요 예)
`Process` 클래스의 `joinCars`,`progressRecords`와 필드가 List 형인데 이 부분을 한번 더 추상화해서 사용할까 생각했습니다.
그런데 과도한 캡슐화로 이용이 어려워지고 이 부분에 대한 행동정의가 어려워서 포기했습니다.
포기하면서 아쉬운 점은 이 필드에 대한 반복문이나 여러 처리가 결국에 `Process` 에서 이루어 지는 것이 만족스럽지 못했습니다. process에 이 기능이 있는것이 바람직하지 못하다고 생각하기 때문입니다.

## 주요 의도
이번에 추가 변경된 각 클래스나 코드 별 어떤 의도인지 말씀드리겠습니다

### `Process`
- 앞서 언급드린 것 처럼 `joinCars`,`progressRecords` 필드를 한번 더 model 로 감싸려고 했는데 과도한 추상화 및 행동 지정이 어려워서 포기했습니다.
- `showResult()` 는 출력을 위한 데이터로 변환화는 것인데요 특히 progressRecords가 List형이다 보니 여기서 어쩔수없이 처리했습니다.
  - Process 에는 적합하지 않다고 들어서 고민이 지금도 있네요 이것을 다른곳으로 옮겨야하나 싶습니다

### `ProgressRecord`
- 가장 큰 목적은 각 횟수마다 게임 기록입니다.
- `simpleResult`은 ResultView 에 전달하는 목적입니다. ResultView 에 게임 진행관련 상태를 전달하고 싶지않아서 리팩터링하면서 추가된 메서드입니다.
  - 그러면서 `simpleResult`가 이미 이 있는데 `joinCarsRecord`의 인자에 Car가 있는채로 Car를 의존해야할까 싶었습니다.(즉 `ProgressRecord` 가 'Car'를 의존해야하는가?) 그럼에도 불구하고`ProgressRecord` 에 model class의 책임은 Car에 대한 위치 기록이 필요하다고 생각해서 유지했습니다.
  - Car의 위치를 단순 정수형으로 변환을 위해 `ProgressRecord`에 `makeSimpleGameResult()` 변환용 메서드를 따로 만들었습니다

### `Rules`
- 인스턴스 생성을 생성자 대신에 `of`를 이용했습니다. 다른 수강생 분들의 리뷰를 보면서 `of` 사용에 대한 고민을 요청주셔서 사용해도 될까 고민 했습니다.
  - 고민 결과 Rules 의 인스턴스 생성은 외부 I/O(API 입력, DB 조회)로 부터 값을 받아서 생성하기 때문에 필요하다고 결정했습니다.
  - 즉 '외부 입력에 대한 검증' 및 '비즈니스 로직으로의 게이트웨이 역할' 표현 가능에 목적을 두었습니다.
- 게임 규칙관련임 처음에는 입력받은 두개 필드(`cars` ,`moves`) 추가하고 , 자동차 전진 기준인 `CAR_FORWARD_CRITERIA` 과 `isProceedByRandom(int randomNum)` 를 추가하였습니다. Random 관련 Util을 모르게 하고자 외부에서 int randomNum 인자로 받아서 작동하게만 했습니다


### `ProcessTest`
- 게임 비즈니스 로직을 테스트 코드 짜면서 방식에 고믾이 많았습니다. 각 게임 입력 상태값 별 결과 출력되는 값이나 상태를 조회하는 것에 어려움이 있었습니다.
  - 그러다보니 탄생한 메서드가 `Process.joinCars()` 입니다. 프로덕션 코드에는 이용되지 않고 테스트 코드를 위해 탄생한 메서드인데 이게 최선인가 고민이 많았습니다.
  - 만약 이 방식에 지양이 권장된다면 어떤 방식으로 접근하면 좋을까요?

## 나가며
요청 주신바대로 적용하면 금방하겠지 하지만, 어제 보다 시간을 쓴 것 같고, 더 많은 고민이 필요한 시간같았습니다.
그러면서 메서드 분리, 클래스 분리의 목적이 무엇인지 계속 고민하게 된 시간같습니다.

좋은 기회 주셔서 감사드립니다 👍

Commit:
add dependency junit-jupiter-params

Mo-bile committed:
Merge branch 'next-step:mo-bile' into mo-bile

Mo-bile committed:
Merge branch 'next-step:mo-bile' into mo-bile

Commit:
docs : 구현할 기능 목록 정리하여
입력, 출력, 진행으로 구분

Commit:
doc : 누락된 기능목록 추가

Commit:
doc : 누락된 기능목록 추가 (자동차 자료형)

Commit:
test : 실패하는 GameProcessTest, CarTest, RandomUtilTest 추가

Commit:
feat : 랜덤 유틸 구현 및 리팩터링 완료

Commit:
test : 랜덤 범위 초과 false 조건 추가 완료

Commit:
feat : car model 개발

Commit:
test : 변경된 car model 기능에 따라 test 변경완료

Commit:
test : 불명확한 기능 및 실행 흐름 다시 재정의하여 테스트 완료

Commit:
refactor : 프로젝트 구조 재정의 및 클래스명 변경하여 유지보수성 및 가독성 향상

Commit:
test : 테스트 수정하여 run() 에 대한 예외를 받을 수있게 변경 완료

Commit:
test : 테스트에 예외처리 에러 메시지 재정의

Commit:
feat : 참가시킬 자동차 수 예외처리 개발

Commit:
feat : 이동횟수 테스트 통과 및 리패터링 완료

Commit:
feat : 참가하는 자동차 기본 세팅 및 리팩터링 완료

Commit:
feat : 게임 이동횟수 기능 개발 및 리팩터링 완료

Commit:
feat : 자동차 이동 기능 개발 및 리팩터링 완료

Commit:
feat : 여러 이동, 여러 자동차 기능 및 리팩터링 완료

Commit:
feat : 기본 입출력 개발

Commit:
fix : car 초기화 변수 0에서 1로 수정

Commit:
refactor
1. 출력문 위치 조절하여 출력문 출력 예상가능하게 함

2. 각 출력문 메서드화

Commit:
refactor : Car doamin에 대한 Init() 메서드를 제거하고 생성자에 추가하여 init() 메서드 호출 실수 방지

Commit:
refactor : DIP 원칙에 따라 추상화에 의존해야 하기 때문에 하위수준인 'ArrayList' 종속 제거

Commit:
test : RacingGameRule 모델 추가하여 레이싱 게임 자체의 rule 에 대한 역할을 분리 및 테스트

Commit:
refactor : RacingGameRules 의 인스턴스 생성 방식을 생성자에서 정적 팩터리 메서드 패턴으로 변경하여, 외부 입력에 대한 검증 및 게이트웨이 역할 표현 가능

Commit:
feat : Rules 모델에 대한 행동 테스트 추가하여 현재 게임 rule 준수 여부 체크 가능

Commit:
refactro : 입력순서에 따라 인자 순서 변경하여 전체 process 내 메서드의 인자와 일치시킴

Commit:
refactor :
1. init() 에 자동차 생성과 게임 이동횟수 제거하여 기본 게임 Rule 세팅에 집중하게 변경 완료
2. run() 에 participateCar() 라는 메서드를 분리하여 입력한 만큼의 자동차 생성방식으로 변경하여 자동차 생성 흐름이 있음을 볼수있게 변경완료
2. cars 를 joinCars로 변경하여 list에대한 목적 명확화

Commit:
feat : 레이싱 기록을 위한 model 을 만들어서 현재 게임 정도 체크와 레이스 기록을 체크 할 수있게 함

Commit:
feat : gameRecord() 인자를 car 단일에서 List로 변경하여 한번에 기록할 수있게끔 변경

Commit:
refactor :
1. progressRecords 에다가 게임을 기록하여 게임 진행현황과 게임 과거 기록을 저장
2. run() 메서드에 게임 결과를 리턴하여 출력 책임을 제거

Commit:
refactor
1. 게임 결과 출력을 기본 타입으로 변경하여 ResultView의 static 메서드인 printAllResult 가 비즈니스 로직의 도메인 및 상태를 모르게 변경
2. 이를 위해 ProgressRecord 에 Car 도메인을 모르는 Interger 형 단순 기록 자로형 추가 및 변환 출력 추가

Commit:
refactor : 랜덤 출력 시 이동 기준을 Rules 도메인으로 이동시켜 메인 Process에서는 게임 흐름에만 역할 집중에게 변경

Commit:
refactor
1. Process 에 주요 메서드 위치 변경
2. ProcessTest 에 일반테스트에서ParameterizedTest 로 변경
3. RulesTest 에 잘못된 매개변수 순서 변경

Commit:
refactor :
1. isProceedByRandom() 메서드 조건 단순화
2. car 생성 시 기본 시작 위치 상수화

javajigi commented on src/main/java/racingGame/Business/Process.java:
```diff
@@ -0,0 +1,66 @@
+    private final List<Car> joinCars = new ArrayList<>();
+    private final List<ProgressRecord> progressRecords = new ArrayList<>();
+    
+    public List<List<Integer>> run(int cars, int moves) {
```
2차원 콜렉션 사용하는 것 지양하면 어떨까요?
2차원 배열 또는 콜렉션을 생기면 `List<Integer>`을 객체로 분리하면 1차원 배열 또는 콜렉션으로 구현할 수 있어요.
이번 기회에 한번 연습해 볼 것을 추천

javajigi commented on src/main/java/racingGame/model/Car.java:
```diff
@@ -0,0 +1,19 @@
+package racingGame.model;
+
+public class Car {
```
Car 객체 분리 👍

javajigi commented on src/main/java/racingGame/model/Car.java:
```diff
@@ -0,0 +1,19 @@
+        this.location = INIT_LOCATION;
+    }
+    
+    public void move() {
```
자동차의 이동 유무에 대한 판단을 Car 객체로 위임하는 것은 어떨까요?
현재 구현은 객체를 너무 수동적인 존재로 바라보고 있는 것 같은데요.
OOP 세계에서 객체는 좀 더 능동적이고, 주도적인 존재로 바라보고 설계하라고 하고 있습니다.

javajigi commented on src/main/java/racingGame/model/Rules.java:
```diff
@@ -0,0 +1,51 @@
+        this.moves = moves;
+    }
+    
+    public static Rules of(int cars, int moves) {
```
굳이 정적 팩터리 메서드로 구현할 필요가 있을까?
이번 기회에 생성자와 정적 팩터리 메서드의 적절한 사용에 대해 고민해 보는 기회를 만들면 어떨까?

AI에게 정적 팩터리 메서드 단점에 대해 질문한 결과

🔍 발견성 (Discoverability) 문제
생성자와의 차이점: 생성자는 자바 API 문서(Javadoc)에서 클래스 이름 아래에 명확하게 표시됩니다. 개발자는 new 키워드를 사용하는 순간 IDE의 자동 완성 기능을 통해 사용 가능한 모든 생성자를 쉽게 확인할 수 있습니다.
정적 팩터리 메서드의 문제: 정적 팩터리 메서드는 일반적인 static 메서드와 동일하게 취급되므로, 개발자가 API 문서나 코드를 직접 살펴보지 않으면 해당 클래스가 객체 생성을 위한 특별한 메서드(팩터리 메서드)를 제공하는지 쉽게 알 수 없습니다.

완화 방법: 이펙티브 자바(Effective Java)에서 제시하는 from, of, valueOf, getInstance 등의 표준 명명 관례를 따르면 이 문제가 완화될 수 있지만, 여전히 개발자가 이 관례를 알고 있어야 합니다.

🚫 상속 제한
필요성: 클래스를 상속(확장)하려면 하위 클래스의 생성자가 상위 클래스의 public 또는 protected 생성자에 접근할 수 있어야 합니다.
정적 팩터리의 영향: 정적 팩터리 메서드를 사용하는 클래스는 객체 생성 로직을 캡슐화하고 재사용 가능한 인스턴스를 관리하기 위해 생성자를 private으로 선언하는 경우가 일반적입니다.

결과: 생성자가 private이면 외부에서 상속이 불가능해집니다. 즉, 정적 팩터리 메서드만 제공하는 클래스는 상속을 통한 확장이 불가능합니다.

참고: 이는 불변(Immutable) 객체를 설계하거나 상속보다 조합(Composition)을 권장하는 설계 원칙에서는 오히려 장점으로 간주되기도 합니다.

🧩 인터페이스 타입으로 구현할 수 없음
자바의 인터페이스는 정적 메서드를 가질 수 있지만 (Java 8 이후), 인터페이스의 인스턴스를 생성하는 정적 팩터리 메서드를 정의하는 것은 불가능합니다.
예외: 자바는 관례적으로 Collections.singletonList()처럼 **별도의 동반 클래스(companion class)**를 만들어 인터페이스의 인스턴스를 반환하는 정적 팩터리 메서드를 제공합니다. (예: java.util.Collections 클래스는 java.util.Collection 인터페이스를 위한 정적 팩터리 메서드들을 모아둔 클래스입니다.)

javajigi commented on src/main/java/racingGame/model/Rules.java:
```diff
@@ -0,0 +1,51 @@
+        return new Rules(cars, moves);
+    }
+    
+    public boolean isLimitMove(int currentMoves) {
```
객체에 메시지를 보내는 OOP적 사고 👍

javajigi commented on src/main/java/racingGame/model/Rules.java:
```diff
@@ -0,0 +1,51 @@
+        return this.moves > currentMoves;
+    }
+    
+    public boolean isLimitCars(int currentCars) {
```
객체에 메시지를 보내는 OOP적 사고 👍

javajigi commented on src/main/java/racingGame/GameApplication.java:
```diff
@@ -0,0 +1,13 @@
+    
+    public static void main(String[] args) {
+        Process process = new Process();
+        ResultView.printAllResult(process.run(InputView.inputCarNum(), InputView.inputMoveNum()));
```
핵심 비지니스 로직을 담당하는 객체와 UI를 잘 분리하고 연결함 👍

javajigi commented on src/test/java/racingGame/Business/ProcessTest.java:
```diff
@@ -0,0 +1,51 @@
+    
+    @ParameterizedTest
+    @ValueSource(ints = {8, 9, 10, 11})
+    void carForwardTest(int moves) {
```
테스트에 대한 복잡도가 높다보니 테스트 구현, 이해 또한 어려움 느낌이 드네요.
이는 Process 객체에 대한 복잡도가 높아 개선이 필요할 수 있다는 힌트가 될 수 있을 것 같아요.
테스트 복잡도를 낮추도록 객체 설계를 개선할 방법은 없을까요?

javajigi reviewed (changes requested):
TDD, OOP가 생각보다 쉽지 않죠?
그럼에도 불구하고 많은 시간을 투자해 테스트 코드를 구현하고 객체 설계를 하고 있는 점이 인상적이네요. 💯 
질문에 대한 답도 아래 남겼는데요.
그 보다 이번 단계에서 생성자 vs 정적 팩터리 메서드 사용에 대해서 고민해 봤으면 좋겠고요.
객체 분리, 과도한 추상화 관련해서는 Process 객체에 개선할 부분이 없는지 고민해 보는 시간을 가진다면 충분히 의미있는 단계가 될 것 같아요.

Q1. 역할과 책임에 집중하여 클래스를 분리하려고 하는데 이부분이 쉽지 않았습니다. 어떤 기준으로 나누는게 좋을까요?

A1. 맥락을 기준으로 객체를 분리하려는 접근 방식은 넘 좋네요. 그런데 설계하는 사람 입장에서 따라 맥락을 구분하는 것이 정성적인 부분이라 쉽지 않은 것 같아요. 맥락(Context)을 변경의 이유로 정의하고, 각 클래스가 오직 그 하나의 이유 때문에 변경되도록 설계하는 연습을 계속해 보세요. 이것이 곧 높은 응집도(Cohesion)와 낮은 결합도(Coupling)를 달성하는 길입니다.

그런데 객체 설계 초보에게는 이 접근 방식이 참 힘들더라고요. 저 또한 그랬고요. 그래서 저는 OOP의 맛을 보기 위해 맥락 기반과 같은 정성적인 방법보다 `객체 지향 생활 체조 원칙`에서 제시하는 원칙 중 클래스 분리(객체 설계)를 강제하는 정량적인 기준이 있는데요. 예를 들어 '규칙 3: 모든 원시값과 문자열을 포장한다.' 원칙입니다. 이런 정량적인 기준을 가지고 객체를 분리하다보면 자연스럽게 OOP의 맛을 느낄 수 있고, 그런 감을 익힌 후에 역할과 책임을 찾기 위한 다음 단계로 도전해 볼 것을 추천합니다.

Q2. 과도한 추상화를 멈추는 기준이 있을까요?
이 또한 정답이 없는 영역이잖아요. 저는 객체 설계 역량을 쌓는 지금 단계에 오히려 과도한 설계를 해보는 것이 시간 낭비가 아닌 좋은 접근 방식이라 생각해요. 우리가 역량을 쌓는 단계는 다음과 같다 생각해요. 너도 과도한 추상화를 해봐야 과도함을 느낄 수 있어 절제하게 되고, 너무 절제하다보니 유지보수성이 떨어진다는 것을 느껴 추상화를 도전하게 되고, 이런 과정이 반복되면서 자신만의 설계 기준이 만들어진다고 봐요. 그렇기 때문에 지금 연습하고 있는 모든 과정이 무의미하다고 생각하지 않습니다. 최근의 경향은 극도의 효율화를 추구하면서 절대로 실수 또는 실패하지 않으려는 경향이 강한 것 같아요. 하지만 정답이 없는 설계 영역의 경우 수 많은 실수와 실패를 경험하는 것이 장기적으로 나의 역량을 강화할 수 있기 때문에 오히려 가장 효율적인 접근 방식이라 생각해요.

Commit:
refactor
1. 랜덤 출력에 대한 결과 값 판단 메서드를 Rules에서 Car 클래스로 이관하여 이동 여부를 Car 가 스스로 판단할 수있게 변경
2. Car 테스트에 Car 인스턴스 생성 setUp() 추가

Commit:
refactor : ProgressRecord에 joinCarsRecord 을 제거하고 simpleResult 을 joinCarsRecord 로 변경하여 클래스 단순화

Commit:
refactor :
1. 2차원 리스트를 방지하기 위한 GameResult 를 만들고 GameResult 를 출력과 리턴에 추가
2. Car 의 이동 연산자를 후위연산자로 변경 및 관련 테스트 보정 하여 + 1 함
3. Car <- ProgressRecord <- GameResult 각 단계별 의존을 명확히 하기위해 메소드 명 변경

Commit:
refacotr :
응답 객체의 불변성 보장을 위해 GameResult를 class 에서 record 로 변경

Commit:
refactor :
of 메서드를 생성자에서 제거

Commit:
refactor : 자동차 생성을 자동차 생성 클래스로 책임을 이관하여 Process의 역할을 줄임

Commit:
refactor : process 의 run() 을 정리하고 메서드 분리한것을 다니 떼낸다.

Commit:
chore : CarFactoryTest 에 필요없는 코드 제거

Mo-bile commented:
안녕하세요 리뷰 주신바를 반영하여 다시 PR드립니다.

무언가 욕심이 생겨서 더 깔끔하게하고 싶은 마음은 커지는데, 리뷰 기회를 놓칠 것 같아서 여기까지 하고 요청 드립니다 👍 

우선 요청 주신 바 대로 
OOP 에서 정성적인 방법(역할, 책임, 맥락) 보다는 객체지향 체조 9원칙이라는 정량적인 원칙을 관점으로 살펴보았습니다.
리뷰 주신 바 대로
1. 2차원 컬렉션을 클래스로 한번 더 감싸기 (규칙 3: 모든 원시값과 문자열을 포장한다.)
2. 참여 자동차 생성을 factory로 이관하여 `Process` 클래스에 역할 줄이기 (규칙 6: 모든 엔티티를 작게 유지한다.)
이렇게 생각하면서 했습니다.

3. 그러면서 ProcessTest 부분의 여러 테스트 중 일부 테스트는 다른 새로 분리된 테스트로 이관되어 테스트의 수 자체는 줄어습니다.
그런데 아직 테스트 코드 작성역량이 부족해서인지 테스트 메서드 내 테스트 코드 작성은 어려움이 있네요. 😂
특히 랜덤으로 출력되는 값에 대한 제어라던지 각각 자동차들의 전진 여부에 대한 예측이라던지 이런 부분이요.

혹시 모호한 부분이 있어 이해가 어려우시면 편하게 말씀 부탁드리겠습니다!
감사합니다 👍

javajigi commented on src/main/java/racingGame/Business/CarFactory.java:
```diff
@@ -0,0 +1,23 @@
+import racingGame.model.Car;
+import racingGame.model.Rules;
+
+public class CarFactory {
```
👍

javajigi commented on src/main/java/racingGame/model/Car.java:
```diff
@@ -0,0 +1,24 @@
+        this.location = INIT_LOCATION;
+    }
+    
+    public int forward() {
```
```suggestion
    public void forward(randomNum) {
         if(isForwardByRandom(randomNum) {
             // 이동
         }
    }
```

forward와 isForwardByRandom을 분리하지 않고 위와 같이 하나로 통합해 구현하는 것은 어떨까?

javajigi commented on src/main/java/racingGame/model/Rules.java:
```diff
@@ -0,0 +1,42 @@
+
+public class Rules {
+    
+    private int cars;
+    private int moves;
```
이 두 개의 int 값이 Rules에 함께 위치하고 있다.
그런데 이 둘의 역할은 달라보인다.
int cars는 자동차 생성 후 불필요한 값이고, moves는 자동차 경주 게임을 진행하는 동안 계속해서 필요한 값이다.
즉, 둘의 생명 주기도 다른데 한 객체에 있다보니 단일 책임의 원칙에도 위반되는 것 같다.
Rules에서 이 두 값을 분리해 보는 것은 어떨까?

Mo-bile replied:
생명주기가 다르다는 중요한점 짚어주셔서 감사합니다!
분리 시도하겠습니다!

javajigi commented on src/main/java/racingGame/ResultView.java:
```diff
@@ -0,0 +1,24 @@
+        for(GameResult result: results) { // 한 게임 단위
+            for(ProgressRecord record: result.progressRecords()) { // 각 라운드
+                for(int location: record.carRecord()) { // 각 자동차의 위치
+                    for(int i = 0; i < location; i++) {
+                        System.out.print("-");
+                    }
```
location에 따라 "-"를 생성하는 부분도 별도의 메서드로 분리하면 테스트도 가능하지 않을까?
`"-".repeat(location)`과 같이 구현하면 테스트도 필요 없을 것으로 보임

javajigi commented on src/test/java/racingGame/Business/CarFactoryTest.java:
```diff
@@ -0,0 +1,22 @@
+    @ParameterizedTest
+    @ValueSource(ints = {3, 4, 5, 6})
+    void carMakeTest(int number) {
+        rules = new Rules(number,1);
+        carFactory = new CarFactory(rules);
```
```suggestion
        List<Car> cars = CarFactory.createCars(number);
```
이와 같이 단순하게 구현할 수 있지 않을까?
Rules까지 참여하다보니 복잡도가 높아 보임

javajigi commented on src/test/java/racingGame/Business/CarFactoryTest.java:
```diff
@@ -0,0 +1,22 @@
+    
+    // 자동차 입력받은만큼 생성했는지?
+    @ParameterizedTest
+    @ValueSource(ints = {3, 4, 5, 6})
```
테스트 데이터가 많다고 무조건 좋은 것은 아니지 않을까?
이런 중복 데이터가 많아지다보면 오히려 전체 테스트 실행 속도도 떨어지고 불필요한 테스트로 보여짐
3과 같이 하나의 경우만 테스트해도 되지 않을까?

javajigi commented on src/test/java/racingGame/model/RulesTest.java:
```diff
@@ -0,0 +1,48 @@
+    @ValueSource(ints = {0, -1, -2, -3})
+    void carSetTest(int num) {
+        assertThatThrownBy(() -> {
+            new Rules(1, num);
```
```suggestion
            new Rules(1, 0);
```
경계 값에 해당하는 0만 테스트해도 되지 않을까?

javajigi reviewed (approved):
피드백 반영하느라 수고했어요. 👍 

코드 리뷰에도 남겼는데요.
좀 더 쉽게 구현할 수 있는 방법이 있을 것 같은데 여러 객체로 분리하고 협력하다보니 오히려 복잡도가 높아지고 그에 따른 테스트 코드의 복잡도 또한 증가하는 느낌이 드네요.
현재 요구사항의 복잡도도 높지 않은데 좀 더 단순하게 구현할 방법을 고민해 보면 어떨까요?
테스트 데이터의 경우 굳이 불필요한 데이터 중복이 많은 것 같아요.
테스트 데이터도 정말 필요한 데이터만 테스트할 수 있도록 의식적인 연습 해볼 것을 추천합니다.

이번 리뷰에서 남긴 피드백은 4단계 진행할 때 함께 반영해 보세요.
4단계 미션 진행할 때 우승자 구하기 부분은 가능하면 TDD로 도전해 볼 것을 추천합니다.
